### PR TITLE
[feat] Add torchscript support for VisualBERT

### PR DESCRIPTION
**Pull request by vedanuj**
- Adds scriptable `VisualBERTForClassification` model and `VisualBERTForPretraining` model
- Adds scriptable versions of some [transformer](https://github.com/huggingface/transformers/blob/v2.3.0/transformers/modeling_bert.py) package BERT model layers : `BertSelfAttentionJit`, `BertAttentionJit`, `BertLayerJit`, `BertEncoderJit`
- Changes to `BertVisioLinguisticEmbeddings` to make it scriptable
- Add tests for `VisualBERT` model

Test Plan:
- Run tests
- Run validation of model on old checkpoints and verify accuracy

### CODE REVIEW

File: mmf/modules/embeddings.py [hunk 1]
```diff
@@ -410,9 +408,9 @@ def forward(
                 )
             else:
                 position_ids_visual = torch.zeros(
-                    *visual_embeddings.size()[:-1],
```
Comment (apsdehal) [inline mmf/modules/embeddings.py:413]: For my information, does `*` causes issues in torch scripting?

File: mmf/modules/hf_layers.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    BertIntermediate,
+    BertOutput,
+    BertSelfAttention,
+    BertSelfOutput,
+)
+
+
+class BertSelfAttentionJit(BertSelfAttention):
+    """
+    Torchscriptable version of `BertSelfAttention` from Huggingface transformers v2.3.0
```
Comment (apsdehal) [inline mmf/modules/hf_layers.py:19]: Is there any source for these modifications or are these novel?

If these are novel, it would good to delineate what exactly and why was changed for future users so they can easily adapt other models for this.

Something very similar has been in works at https://github.com/huggingface/transformers/pull/6846

**Review by apsdehal — COMMENTED** [review]
Thanks for this awesome feature. Leaving some comments.

File: mmf/modules/hf_layers.py (hunk 2 above)
Comment (vedanuj) [reply mmf/modules/hf_layers.py:19]: These are novel modifications. I will update in the comment what has changed for each of these. The huggingface PR isn't merging it seems.

File: mmf/modules/embeddings.py (hunk 1 above)
Comment (vedanuj) [reply mmf/modules/embeddings.py:413]: Static size inference is not yet possible with torchscript.

File: mmf/modules/embeddings.py [hunk 3]
```diff
@@ -362,36 +362,34 @@ def forward(
...
                 # Get rid of the -1.
                 image_text_alignment = image_text_alignment_mask * image_text_alignment
 
                 # position_embeddings_visual
                 # = Batch x image_length x alignment length x dim
                 position_embeddings_visual = self.position_embeddings(
                     image_text_alignment
-                ) * image_text_alignment_mask.to(
-                    dtype=next(self.parameters()).dtype
-                ).unsqueeze(
```
[the remark this answers is not in the document]
Comment (vedanuj) [reply mmf/modules/embeddings.py:375]: Tested on CPU. CPU runs are passing on Actions. CircleCI has some memory problem, will need to check that later.

File: mmf/modules/embeddings.py [hunk 4]
```diff
@@ -362,36 +362,34 @@ def forward(
...
-                image_text_alignment_mask = image_text_alignment_mask.to(
-                    dtype=next(self.parameters()).dtype
-                ).sum(2)
+                image_text_alignment_mask = image_text_alignment_mask.sum(2)
                 image_text_alignment_mask[
                     image_text_alignment_mask == 0
-                ] = 1  # Avoid devide by zero error
+                ] = torch.tensor(
+                    [1]
+                )  # Avoid devide by zero error
```
[the remark this answers is not in the document]
Comment (vedanuj) [reply mmf/modules/embeddings.py:384]: Sorry I didn't understand this comment.

File: mmf/models/visual_bert.py [hunk 5]
```diff
@@ -115,29 +115,24 @@ def forward(
...
-                embedding_output, extended_attention_mask, self.fixed_head_masks
-            )
+            encoded_layers = self.encoder(embedding_output, extended_attention_mask)
             sequence_output = encoded_layers[0]
             pooled_output = self.pooler(sequence_output)
-            attn_data_list = []
+            attn_data_list: List[Tensor] = []
 
-            if self.output_attentions:
+            if not torch.jit.is_scripting() and self.output_attentions:
```
[the remark this answers is not in the document]
Comment (vedanuj) [reply mmf/models/visual_bert.py:134]: Sounds good. I will make these changes.

**Review by apsdehal — APPROVED** [review]
